### ISSUE TITLE
DOC: Document einsum's index parsing and mapping code

### ISSUE THREAD

**Issue by jaimefrio** [numpy/numpy#10801]
The code parsing and mapping indices in einsum is cryptic and hard to follow. We should add meaningful descriptions and examples of how indices are parsed and how the information is stored and processed before actually doing any sums or products on the values.

**Comment by rsokl** [comment]
Hi guys, I would also recommend improving the documentation for keyword argument `optimize`. Presently, the documentation suggests that `optimize` is used for determining an optimal path for performing intermediate contractions. While this is the case, it fails to mention that this also enables the use of BLAS where possible! This is relevant even for computations where intermediate contractions are not important.
